File: LLD/AmazonSystem.java
package com.amazon.lld;

import java.util.*;

/*
 * ==============================================================================================
 * AMAZON LOW LEVEL DESIGN (LLD) - INTERVIEW REFERENCE
 * ==============================================================================================
 * 
 * Key Design Patterns Used:
 * 1. Singleton Pattern: Ensures valid global access to the Catalog (inventory).
 * 2. Builder Pattern: Manages complexity of creating Product objects with many optional fields.
 * 3. Strategy Pattern: Allows interchangeable algorithms for Payment and Search behavior.
 * 4. Factory Pattern (Implicit): Could be used for creating different Account types (omitted for brevity).
 *
 * SOLID Principles Applied:
 * - SRP (Single Responsibility): Each class (Order, Cart, Product) handles one domain aspect.
 * - OCP (Open/Closed): PaymentStrategy allows adding new payment methods without changing Order logic.
 * - LSP (Liskov Substitution): Customer/Admin can substitute the abstract Account class.
 * - ISP (Interface Segregation): Interfaces are focused (SearchService, PaymentStrategy).
 * - DIP (Dependency Inversion): High-level modules depend on abstractions (PaymentStrategy interface).
 */

public class AmazonSystem {
    public static void main(String[] args) {
        System.out.println("--- Amazon Low Level Design Demo ---");

        // SYSTEM INITIALIZATION
        // Singleton Access: Ensures only one instance of the catalog exists.
        Catalog catalog = Catalog.getInstance();
        NotificationService notificationService = new NotificationService();

        // -----------------------------------------------------
        // 1. ADMIN FLOW: ADDING PRODUCTS
        // -----------------------------------------------------
        Address adminAddr = new Address("HQ", "Seattle", "WA", "98109", "USA");
        Admin admin = new Admin("admin", "pass", "admin@amazon.com", "123", adminAddr);

        // PATTERN: BUILDER
        // Why? Product creation involves many optional parameters (desc, rating, etc.).
        // A constructor with 10 arguments is hard to read and sensitive to order errors.
        Product laptop = new Product.ProductBuilder("MacBook Pro", "P001")
                            .description("Apple MacBook Pro 16 inch")
                            .price(2500.00)
                            .availableItemCount(10)
                            .category(new ProductCategory("Electronics", "Gadgets"))
                            .build();

        admin.addProduct(laptop);
        catalog.addProduct(laptop); // Adding to global inventory

        // -----------------------------------------------------
        // 2. CUSTOMER FLOW: SEARCH & CART
        // -----------------------------------------------------
        Address custAddr = new Address("123 Main St", "New York", "NY", "10001", "USA");
        Customer customer = new Customer("john_doe", "passwd", "john@example.com", "555-0100", custAddr);

        // PATTERN: STRATEGY (in SearchService)
        // Catalog implements SearchService, allowing us to swap search logic (BS Tree, Hash Map, External Engine)
        // without affecting the client code.
        List<Product> searchResults = catalog.search("MacBook Pro");
        if (searchResults.isEmpty()) {
            System.out.println("Product not found");
            return;
        }
        Product productToBuy = searchResults.get(0);
        System.out.println("Customer found product: " + productToBuy.getName());

        Item cartItem = new Item("I001", 1, productToBuy.getPrice());
        customer.getShoppingCart().addItem(cartItem);
        System.out.println("Item added to cart.");

        // -----------------------------------------------------
        // 3. ORDER PLACEMENT
        // -----------------------------------------------------
        List<Item> orderItems = new ArrayList<>(customer.getShoppingCart().getItems());
        Order order = new Order("O-1001", orderItems, 2500.00);
        customer.addOrder(order);
        System.out.println("Order placed: " + order.getOrderNumber());

        // -----------------------------------------------------
        // 4. PAYMENT & NOTIFICATION
        // -----------------------------------------------------
        
        // PATTERN: STRATEGY (Payment)
        // We can easily switch this to new PaypalPayment(...) without changing the logic below.
        PaymentStrategy payment = new CreditCardPayment("4111-1111", "123");
        
        if(payment.processPayment(2500.00)) {
            // State Transition: Order moves to PENDING/CONFIRMED
            order.setStatus(OrderStatus.PENDING);
            System.out.println("Payment Successful via Credit Card.");
        }

        // Notification triggers (could be Observer pattern in a more complex setup where 
        // User 'subscribes' to OrderStatus changes)
        notificationService.sendOrderConfirmation(order, customer);

        // -----------------------------------------------------
        // 5. SHIPPING
        // -----------------------------------------------------
        Shipment shipment = new Shipment("S-999", new Date(), "FedEx");
        System.out.println("Shipment created with status: " + shipment.getStatus());
    }
}

// ==========================================
// CORE ENUMS (State Management)
// ==========================================

enum OrderStatus {
    UNSHIPPED, PENDING, SHIPPED, COMPLETED, CANCELED, REFUND_APPLIED
}

enum PaymentStatus {
    UNPAID, PENDING, COMPLETED, FILLED, DECLINED, CANCELLED, REFUNDED
}

enum ShipmentStatus {
    PENDING, SHIPPED, DELIVERED, ON_HOLD
}

enum AccountStatus {
    ACTIVE, BLOCKED, BANNED, COMPROMISED, ARCHIVED, UNKNOWN
}

// ==========================================
// DOMAIN MODELS
// ==========================================

// Value Object: Address is immutable in practice (though here fields are simplified)
class Address {
    private String streetAddress;
    private String city;
    private String state;
    private String zipCode;
    private String country;

    public Address(String streetAddress, String city, String state, String zipCode, String country) {
        this.streetAddress = streetAddress;
        this.city = city;
        this.state = state;
        this.zipCode = zipCode;
        this.country = country;
    }

    @Override
    public String toString() {
        return streetAddress + ", " + city + ", " + state + " " + zipCode + ", " + country;
    }
}

class ProductCategory {
    private String name;
    private String description;

    public ProductCategory(String name, String description) {
        this.name = name;
        this.description = description;
    }

    public String getName() { return name; }
}

/**
 * Product Class
 * Uses Builder Pattern to handle complex object construction.
 * Implements immutability effectively by using private constructor.
 */
class Product {
    private String productID;
    private String name;
    private String description;
    private double price;
    private ProductCategory category;
    private int availableItemCount;

    // Private constructor forces use of Builder
    private Product(ProductBuilder builder) {
        this.productID = builder.productID;
        this.name = builder.name;
        this.description = builder.description;
        this.price = builder.price;
        this.category = builder.category;
        this.availableItemCount = builder.availableItemCount;
    }

    public String getName() { return name; }
    public double getPrice() { return price; }
    public ProductCategory getCategory() { return category; }

    // Static Inner Builder Class
    public static class ProductBuilder {
        // Mandatory fields
        private String productID;
        private String name;
        
        // Optional fields
        private String description;
        private double price;
        private ProductCategory category;
        private int availableItemCount;

        public ProductBuilder(String name, String productID) {
            this.name = name;
            this.productID = productID;
        }

        public ProductBuilder description(String description) {
            this.description = description;
            return this;
        }
        public ProductBuilder price(double price) {
            this.price = price;
            return this;
        }
        public ProductBuilder category(ProductCategory category) {
            this.category = category;
            return this;
        }
        public ProductBuilder availableItemCount(int count) {
            this.availableItemCount = count;
            return this;
        }
        public Product build() {
            return new Product(this);
        }
    }
}

class Item {
    private String itemID;
    private int quantity;
    private double price;

    public Item(String itemID, int quantity, double price) {
        this.itemID = itemID;
        this.quantity = quantity;
        this.price = price;
    }

    public void updateQuantity(int quantity) {
        this.quantity = quantity;
    }
}

class ShoppingCart {
    private List<Item> items;

    public ShoppingCart() {
        this.items = new ArrayList<>();
    }

    public void addItem(Item item) {
        this.items.add(item);
    }
    
    public List<Item> getItems() {
        return items;
    }
}

class Order {
    private String orderNumber;
    private OrderStatus status;
    private Date orderDate;
    private List<Item> items;
    private double totalAmount;

    public Order(String orderNumber, List<Item> items, double totalAmount) {
        this.orderNumber = orderNumber;
        this.items = items;
        this.totalAmount = totalAmount;
        this.orderDate = new Date();
        this.status = OrderStatus.PENDING;
    }

    public void setStatus(OrderStatus status) {
        this.status = status;
        // In a real system, setting status might notify observers (User, Warehouse)
    }
    
    public String getOrderNumber() { return orderNumber; }
    
    @Override
    public String toString() {
        return "Order #" + orderNumber;
    }
}

class Shipment {
    private String shipmentNumber;
    private Date shipmentDate;
    private String shipmentMethod;
    private ShipmentStatus status;

    public Shipment(String shipmentNumber, Date shipmentDate, String shipmentMethod) {
        this.shipmentNumber = shipmentNumber;
        this.shipmentDate = shipmentDate;
        this.shipmentMethod = shipmentMethod;
        this.status = ShipmentStatus.PENDING;
    }

    public ShipmentStatus getStatus() { return status; }
}

/**
 * Account - Abstract Base Class
 * Supports Open/Closed Principle. New account types (e.g. Vendor, PrimeMember) 
 * can be added by extending this class without modifying existing code.
 */
abstract class Account {
    private String userName;
    private String password;
    private AccountStatus status;
    protected String email;
    private String phone;
    private Address shippingAddress;

    public Account(String userName, String password, String email, String phone, Address address) {
        this.userName = userName;
        this.password = password;
        this.email = email;
        this.phone = phone;
        this.shippingAddress = address;
        this.status = AccountStatus.ACTIVE;
    }

    public String getEmail() { return email; }
}

class Customer extends Account {
    private ShoppingCart cart;
    private List<Order> orders;

    public Customer(String userName, String password, String email, String phone, Address address) {
        super(userName, password, email, phone, address);
        this.cart = new ShoppingCart();
        this.orders = new ArrayList<>();
    }

    public ShoppingCart getShoppingCart() { return cart; }
    public void addOrder(Order order) { orders.add(order); }
}

class Admin extends Account {
    public Admin(String userName, String password, String email, String phone, Address address) {
        super(userName, password, email, phone, address);
    }

    public void addProduct(Product product) {
        // Admin specific logic for adding products
        System.out.println("Admin adding product: " + product.getName());
    }
}

// ==========================================
// SERVICES & INTERFACES
// ==========================================

/**
 * Interface Segregation Principle:
 * Clients depend on specific interfaces (SearchService) rather than a monolithic class.
 */
interface SearchService {
    List<Product> search(String query);
}

/**
 * Singleton Pattern
 * Why? We need a single source of truth for the Product Catalog in memory.
 * Thread Safety: 'synchronized' ensures two threads don't create duplicate instances.
 */
class Catalog implements SearchService {
    private static Catalog instance;
    private Map<String, List<Product>> productNames;

    // Private constructor prevents direct instantiation
    private Catalog() {
        productNames = new HashMap<>();
    }

    public static synchronized Catalog getInstance() {
        if (instance == null) {
            instance = new Catalog();
        }
        return instance;
    }

    public void addProduct(Product product) {
        productNames.computeIfAbsent(product.getName(), k -> new ArrayList<>()).add(product);
    }

    @Override
    public List<Product> search(String query) {
        return productNames.containsKey(query) ? productNames.get(query) : new ArrayList<>();
    }
}

/**
 * Strategy Pattern
 * Encapsulates a family of algorithms (Payment methods) and makes them interchangeable.
 * The Client (Order processing) doesn't need to know the details of Paypal vs CreditCard.
 */
interface PaymentStrategy {
    boolean processPayment(double amount);
}

class CreditCardPayment implements PaymentStrategy {
    private String cardNumber;
    private String cvv;

    public CreditCardPayment(String cardNumber, String cvv) {
        this.cardNumber = cardNumber;
        this.cvv = cvv;
    }

    @Override
    public boolean processPayment(double amount) {
        System.out.println("Processing Credit Card Payment of $" + amount);
        return true;
    }
}

class NotificationService {
    public void sendOrderConfirmation(Order order, Account account) {
        System.out.println("Email sent to " + account.getEmail() + " for Order " + order);
    }
}

File: LLD/AmazonSystem.py
# fmt: off
# ==============================================================================
#  AMAZON SYSTEM — ASCII CLASS DIAGRAM
# ==============================================================================
#
#  ┌──────────────────────────────────────────────────────────────────────────┐
#  │                        AMAZON SYSTEM (e-commerce)                        │
#  └──────────────────────────────────────────────────────────────────────────┘
#
#  ┌──────────────────────────────┐    ┌─────────────────────────────────┐
#  │          Catalog             │    │           Account (ABC)         │
#  │      (Singleton)             │    ├─────────────────────────────────┤
#  ├──────────────────────────────┤    │ + username: str                 │
#  │ + _instance: Catalog         │    │ + email: str                    │
#  │ + products: Dict[id,Product] │    │ + address: str                  │
#  ├──────────────────────────────┤    └────────────┬────────────────────┘
#  │ + get_instance(): Catalog    │                 │
#  │ + add_product(product)       │        ┌────────┴────────┐
#  │ + search(query): List[Prod]  │        │                 │
#  └──────────────────────────────┘        ▼                 ▼
#      ← implements SearchService →  ┌──────────┐     ┌───────────┐
#                                    │ Customer │     │   Admin   │
#  ┌──────────────────────────────┐  ├──────────┤     ├───────────┤
#  │          Product             │  │ + cart:  │     │ + add_    │
#  ├──────────────────────────────┤  │  Shopping│     │  product()│
#  │ + product_id: str            │  │  Cart    │     └───────────┘
#  │ + name: str                  │  │ + orders │
#  │ + description: str           │  │   : List │
#  │ + price: float               │  ├──────────┤
#  │ + category: str              │  │+add_order│
#  ├──────────────────────────────┤  └────┬─────┘
#  │ Builder (inner class)        │       │ 1 owns
#  │  + set_name() / set_price()  │       ▼
#  │  + set_description()         │  ┌───────────────────────────────┐
#  │  + build(): Product          │  │         ShoppingCart          │
#  └──────────────────────────────┘  ├───────────────────────────────┤
#                                    │ + items: List[Item]            │
#  ┌──────────────────────────────┐  ├───────────────────────────────┤
#  │          Order               │  │ + add_item(product, qty)       │
#  ├──────────────────────────────┤  │ + remove_item(product_id)      │
#  │ + order_id: str              │  │ + total(): float               │
#  │ + items: List[Item]          │  └───────────────────────────────┘
#  │ + total_amount: float        │
#  │ + status: OrderStatus (enum) │  ┌──────────────────────────────┐
#  ├──────────────────────────────┤  │     PaymentStrategy (ABC)    │
#  │ + set_status()               │  ├──────────────────────────────┤
#  └──────────────────────────────┘  │ + process(amount): bool      │
#                                    └──────────────┬───────────────┘
#  ┌──────────────────────────────┐                 │
#  │            Item              │      ┌──────────┴──────────┐
#  ├──────────────────────────────┤      ▼                     ▼
#  │ + product: Product           │  ┌──────────────┐  ┌────────────────┐
#  │ + quantity: int              │  │ CreditCard   │  │PayPalPayment   │
#  │ + price: float               │  │ Payment      │  └────────────────┘
#  └──────────────────────────────┘  └──────────────┘
#
#  ┌──────────────────────────────┐   ┌──────────────────────────────┐
#  │  NotificationService         │   │     SearchService (ABC)      │
#  ├──────────────────────────────┤   ├──────────────────────────────┤
#  │ + send_email_notification()  │   │ + search(query): List[Prod]  │
#  │ + send_shipment_update()     │   └──────────────────────────────┘
#  │ + create_shipment()          │
#  └──────────────────────────────┘
#
#  RELATIONSHIPS:
#  Catalog (Singleton) ──*──> Product      (manages all products, SearchService)
#  Customer ──▷── Account                  (inherits)
#  Admin    ──▷── Account                  (inherits)
#  Customer ──1──> ShoppingCart            (owns a cart)
#  Customer ──*──> Order                   (places orders)
#  Order ──*──> Item                       (line items per order)
#  Item ──1──> Product                     (references product)
#  Product uses Builder Pattern            (fluent construction)
#  CreditCardPayment / PayPalPayment ──▷── PaymentStrategy (implements)
# ==============================================================================
# fmt: on
from abc import ABC, abstractmethod
from enum import Enum, auto
from datetime import datetime
from typing import List, Dict, Optional

"""
==============================================================================================
AMAZON LOW LEVEL DESIGN (LLD) - PYTHON IMPLEMENTATION
==============================================================================================

Key Design Patterns:
1. Singleton: Thread-safe Catalog instance.
2. Builder: Used for Product creation (explicitly implemented to demonstrate pattern).
3. Strategy: PaymentStrategy for interchangeable payment methods.
4. Observer (Implicit): Notification service reacting to events.

Class Design Diagram:
---------------------
[Account] <|-- [Customer]
[Account] <|-- [Admin]
[Customer] "1" *-- "1" [ShoppingCart]
[Customer] "1" *-- "0..*" [Order]
[Order] "1" *-- "1..*" [Item]
[Catalog] ..> [Product] : Manages
[SearchService] <|.. [Catalog] : Implements
[PaymentStrategy] <|.. [CreditCardPayment] : Implements

Class Details:
---------------------
1. Product
   - Attributes: product_id, name, description, price, category.
   - Methods: Builder (inner class) -> build().

2. Catalog (Singleton)
   - Role: Inventory management.
   - Methods: add_product(), search() [SearchService impl].

3. Account (ABC)
   - Role: Base User class.
   - Attributes: username, email, address.

4. Customer
   - Attributes: cart (ShoppingCart).
   - Methods: add_order().

5. Order
   - Role: Transaction record.
   - Attributes: items, total_amount, status.
   - Methods: set_status().

6. PaymentStrategy (ABC)
   - Methods: process_payment(amount).
"""

# ==========================================
# ENUMS
# ==========================================

class OrderStatus(Enum):
    UNSHIPPED = auto()
    PENDING = auto()
    SHIPPED = auto()
    COMPLETED = auto()
    CANCELED = auto()
    REFUND_APPLIED = auto()

class PaymentStatus(Enum):
    UNPAID = auto()
    PENDING = auto()
    COMPLETED = auto()
    DECLINED = auto()
    CANCELLED = auto()
    REFUNDED = auto()

class ShipmentStatus(Enum):
    PENDING = auto()
    SHIPPED = auto()
    DELIVERED = auto()
    ON_HOLD = auto()

class AccountStatus(Enum):
    ACTIVE = auto()
    BLOCKED = auto()
    BANNED = auto()
    UNKNOWN = auto()

# ==========================================
# DOMAIN MODELS
# ==========================================

class Address:
    def __init__(self, street: str, city: str, state: str, zip_code: str, country: str):
        self.street = street
        self.city = city
        self.state = state
        self.zip_code = zip_code
        self.country = country

    def __str__(self):
        return f"{self.street}, {self.city}, {self.state} {self.zip_code}, {self.country}"

class ProductCategory:
    def __init__(self, name: str, description: str):
        self.name = name
        self.description = description

class Product:
    """
    Product Class
    Demonstrates the Builder Pattern.
    """
    def __init__(self, product_id: str, name: str, description: str, 
                 price: float, category: ProductCategory, available_count: int):
        self._product_id = product_id
        self._name = name
        self._description = description
        self._price = price
        self._category = category
        self._available_count = available_count

    @property
    def name(self):
        return self._name

    @property
    def price(self):
        return self._price

    @property
    def category(self):
        return self._category

    class Builder:
        """
        Inner Builder Class to handle complex object construction.
        In Python, named arguments often suffice, but this pattern is 
        useful when construction logic is complex or invariant checking is needed.
        """
        def __init__(self, name: str, product_id: str):
            self._name = name
            self._product_id = product_id
            self._description = ""
            self._price = 0.0
            self._category = None
            self._available_count = 0

        def set_description(self, description: str):
            self._description = description
            return self

        def set_price(self, price: float):
            self._price = price
            return self

        def set_category(self, category: ProductCategory):
            self._category = category
            return self

        def set_available_count(self, count: int):
            self._available_count = count
            return self

        def build(self):
            return Product(
                self._product_id, self._name, self._description,
                self._price, self._category, self._available_count
            )

class Item:
    def __init__(self, item_id: str, quantity: int, price: float):
        self.item_id = item_id
        self.quantity = quantity
        self.price = price

    def update_quantity(self, quantity: int):
        self.quantity = quantity

class ShoppingCart:
    def __init__(self):
        self.items: List[Item] = []

    def add_item(self, item: Item):
        self.items.append(item)

    def get_items(self) -> List[Item]:
        return self.items

class Order:
    def __init__(self, order_number: str, items: List[Item], total_amount: float):
        self.order_number = order_number
        self.items = items
        self.total_amount = total_amount
        self.order_date = datetime.now()
        self.status = OrderStatus.PENDING

    def set_status(self, status: OrderStatus):
        self.status = status
    
    def __str__(self):
        return f"Order #{self.order_number}"

class Shipment:
    def __init__(self, shipment_number: str, shipment_method: str):
        self.shipment_number = shipment_number
        self.shipment_date = datetime.now()
        self.shipment_method = shipment_method
        self.status = ShipmentStatus.PENDING

# ==========================================
# ACCOUNTS
# ==========================================

class Account(ABC):
    def __init__(self, username: str, password: str, email: str, phone: str, address: Address):
        self.username = username
        self.password = password
        self.email = email
        self.phone = phone
        self.shipping_address = address
        self.status = AccountStatus.ACTIVE

class Customer(Account):
    def __init__(self, username: str, password: str, email: str, phone: str, address: Address):
        super().__init__(username, password, email, phone, address)
        self.cart = ShoppingCart()
        self.orders: List[Order] = []

    def add_order(self, order: Order):
        self.orders.append(order)

class Admin(Account):
    def add_product(self, product: Product):
        print(f"Admin adding product: {product.name}")

# ==========================================
# SERVICES & STRATEGIES
# ==========================================

class SearchService(ABC):
    @abstractmethod
    def search(self, query: str) -> List[Product]:
        pass

class Catalog(SearchService):
    """
    Singleton Catalog
    Manages the products in the system.
    """
    _instance = None
    _product_names: Dict[str, List[Product]] = {}

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(Catalog, cls).__new__(cls)
            cls._product_names = {}
        return cls._instance

    def add_product(self, product: Product):
        if product.name not in self._product_names:
            self._product_names[product.name] = []
        self._product_names[product.name].append(product)

    def search(self, query: str) -> List[Product]:
        # Fast lookup by name
        return self._product_names.get(query, [])

class PaymentStrategy(ABC):
    @abstractmethod
    def process_payment(self, amount: float) -> bool:
        pass

class CreditCardPayment(PaymentStrategy):
    def __init__(self, card_number: str, cvv: str):
        self.card_number = card_number
        self.cvv = cvv

    def process_payment(self, amount: float) -> bool:
        print(f"Processing Credit Card Payment of ${amount:.2f}")
        # Logic to contact bank API would go here
        return True

class NotificationService:
    def send_order_confirmation(self, order: Order, account: Account):
        print(f"Email sent to {account.email} for {order}")

# ==========================================
# MAIN DEMO
# ==========================================

if __name__ == "__main__":
    print("--- Amazon Low Level Design Demo (Python) ---")

    # 1. Initialize System Elements (Singleton Catalog)
    catalog = Catalog()
    notification_service = NotificationService()

    # 2. Admin adds products
    admin_addr = Address("HQ", "Seattle", "WA", "98109", "USA")
    admin = Admin("admin", "pass", "admin@amazon.com", "123", admin_addr)

    # Creating Product using Builder Pattern
    laptop = (Product.Builder("MacBook Pro", "P001")
              .set_description("Apple MacBook Pro 16 inch")
              .set_price(2500.00)
              .set_category(ProductCategory("Electronics", "Gadgets"))
              .set_available_count(10)
              .build())

    admin.add_product(laptop)
    catalog.add_product(laptop)

    # 3. Customer Actions
    cust_addr = Address("123 Main St", "New York", "NY", "10001", "USA")
    customer = Customer("john_doe", "passwd", "john@example.com", "555-0100", cust_addr)

    # 4. Search (Strategy)
    search_results = catalog.search("MacBook Pro")
    if not search_results:
        print("Product not found")
        exit()
    
    product_to_buy = search_results[0]
    print(f"Customer found product: {product_to_buy.name}")

    # 5. Add to Cart
    cart_item = Item("I001", 1, product_to_buy.price)
    customer.cart.add_item(cart_item)
    print("Item added to cart.")

    # 6. Checkout
    order_items = list(customer.cart.get_items()) # Create copy
    order = Order("O-1001", order_items, 2500.00)
    customer.add_order(order)
    print(f"Order placed: {order.order_number}")

    # 7. Payment (Strategy Usage)
    payment = CreditCardPayment("4111-1234", "999")
    if payment.process_payment(2500.00):
        order.set_status(OrderStatus.PENDING)
        print("Payment Successful via Credit Card.")

    # 8. Notification
    notification_service.send_order_confirmation(order, customer)

    # 9. Shipment
    shipment = Shipment("S-999", "FedEx")
    print(f"Shipment created with status: {shipment.status.name}")
